mm2src/lp_ordermatch.rs

```diff
@@ -5282,3 +5431,61 @@ fn choose_taker_confs_and_notas(
         taker_coin_nota,
     }
 }
+
+#[derive(Debug, Serialize)]
+#[serde(tag = "address_type", content = "address_data")]
+pub enum OrderbookAddress {
+    Transparent(String),
+    Shielded,
+}
+
+#[derive(Debug, Display)]
+enum OrderbookAddrErr {
+    AddrFromPubkeyError(String),
+    CoinIsNotSupported(String),
+    DeserializationError(json::Error),
+    InvalidPlatformCoinProtocol(String),
+    PlatformCoinConfIsNull(String),
+}
+
+impl From<json::Error> for OrderbookAddrErr {
+    fn from(err: json::Error) -> Self { OrderbookAddrErr::DeserializationError(err) }
+}
+
+fn orderbook_address(
```

Comment:
How about moving this function to the `coins` crate? It may return not `OrderbookAddress`, but another type with the same variants. I think this is the `coins` crate specific implementation, and `ordermatch` shouldn't know about the details like `CoinProtocol`, parsing addresses from pubkeys etc.

Reply:
That's quite a controversial question :slightly_smiling_face: This way coins crate will know about `OrderbookAddress` and we will get a kind of cycle dependency: `ordermatch` depends on `coins` and `coins` implicitly depends on implementation details of `ordermatch`. My long-term goal is to avoid making such dependencies and, even more, remove traits like `SwaOps` from `coins`.

> and ordermatch shouldn't know about the details like CoinProtocol, parsing addresses from pubkeys etc

 `CoinProtocol` is a public struct that is used by other crates in our dependency tree. I don't see any problems in using in `ordermatch` too.

Reply:
I forgot to mention that such function in `coins` crate may return something not related to ordermatching protocol
> It may return not OrderbookAddress, but another type with the same variants

For example, it may be called `market_address` and it may be a part of `MarketCoinOps`.

Reply:
This code can be easy moved to `coins` crate. What I'd like is to avoid code that only one crate needs to know about.
I think you agree that `ordermatch` crate shouldn't know about kind/implementation of coins like `UTXO`, `ETH`, `ZCOIN` etc. It should interact with the coins through the traits like `MmCoin`, `MarketCoin`.

Reply:
I see it a bit differently: `coins` encapsulates and abstracts over the functionality like `MarketCoinOps` and other traits. `MmCoinEnum` itself is public, and specific coin types are already used in `coins_activation`. 

> This code can be easily moved to coins crate. What I'd like is to avoid code that only one crate needs to know about.

Why `coins` should know about the stuff that is *now* used only in the orderbook 2.0 RPC?

Reply:
> What I'd like is to avoid code that only one crate needs to know about.

As of now, only crate knows about `OrderbookAddress`. If we move this to `coins`, we will have *two* crates knowing about this struct. `CoinProtocol` and other stuff is already public, so *everyone* knows about it.

Even more, if we consider a *possible* future usage (which we can't predict - if I could predict the future I would buy bitcoin in 2012 🙂 ) and make `market_address`, and then start using it somewhere else, we can end up adding the variant to `MarketAddress`, which is *not* needed in the `orderbook` 2.0 RPC. We will have to make a hard decision on how to handle this unneeded variant.